### PR TITLE
core/services/ccipcapability: add the ccip capability package

### PR DESCRIPTION
**Pull request by makramkd**
## Architecture
ccip-init-mini

### CODE REVIEW

File: core/services/ccipcapability/types/types.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	Capabilities []keystone_capability_registry.CapabilityRegistryCapability
+	Nodes        []keystone_capability_registry.CapabilityRegistryNodeInfo
+}
+
+type CapabilityRegistry interface {
+	// LatestState returns the latest state of the on-chain capability registry.
+	LatestState() (RegistryState, error)
+}
+
+type ChainConfig interface {
```
Comment (asoliman92) [inline core/services/ccipcapability/types/types.go:24]: This is different on the contract, it includes ChainSelector as well.

File: core/services/ccipcapability/delegate.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	lggr            logger.Logger
+	registrarConfig plugins.RegistrarConfig
+	pipelineRunner  pipeline.Runner
+	relayGetter     RelayGetter
+	capRegistry     cctypes.CapabilityRegistry
+	keystore        keystore.Master
+	ds              sqlutil.DataSource
+	peerWrapper     *ocrcommon.SingletonPeerWrapper
+
+	isNewlyCreatedJob bool
```
Comment (winder) [inline core/services/ccipcapability/delegate.go:41]: What do you imagine this is going to be used for? It's currently passed into the oracle creator which effectively makes it a constant.

File: core/services/ccipcapability/launcher/bluegreen.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package launcher
+
+import (
+	"go.uber.org/multierr"
+
+	cctypes "github.com/smartcontractkit/chainlink/v2/core/services/ccipcapability/types"
+)
+
+// blueGreenDeployment represents a blue-green deployment of OCR instances.
+type blueGreenDeployment struct {
```
Comment (winder) [inline core/services/ccipcapability/launcher/bluegreen.go:12]: Where are capabilities actually deployed? Seems weird to have this built into the node.

Currently it seems like this only manages switching between configs. It might be nice to switch between LOOPPs as well.

File: core/services/ccipcapability/launcher/launcher.go [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+			don.Id, err)
+	}
+
+	// valid cases:
+	// a) len(commitOCRConfigs) == 2 && ceDep.NumCommitInstances() == 1: this is a new green instance.
+	// b) len(commitOCRConfigs) == 1 && ceDep.NumCommitInstances() == 2: this is a promotion of green->blue.
+	// invalid cases (enforced in the config contract):
+	// a) len(commitOCRConfigs) == 2 && ceDep.NumCommitInstances() == 2: this is an invariant violation.
+	// b) len(commitOCRConfigs) == 1 && ceDep.NumCommitInstances() == 1: this is an invariant violation.
+	// same thing applies to exec.
```
Comment (winder) [inline core/services/ccipcapability/launcher/launcher.go]: This seems like the sort of thing that could be pushed into the bluegreen helper so you can do something like
```
if err := ceDeps.commit.Update(commitConfigs); err != nil {
	return fmt.Errorf("failed to update commit deployment: %w", err)
}

if err := ceDeps.execute.Update(executeConfigs); err != nil {
	return fmt.Errorf("failed to update execute deployment: %w", err)
}
```

File: core/services/job/models.go [hunk 5]
```diff
@@ -50,6 +50,7 @@ const (
 	Webhook                 Type = (Type)(pipeline.WebhookJobType)
 	Workflow                Type = (Type)(pipeline.WorkflowJobType)
 	StandardCapabilities    Type = (Type)(pipeline.StandardCapabilitiesJobType)
+	CCIP                    Type = (Type)(pipeline.CCIPJobType)
```
Comment (winder) [inline core/services/job/models.go:53]: It appears that someone was trying to alphabetize these, maybe move this to line 39. Same with the flags below.

**Review by winder — COMMENTED** [review]
I haven't spent enough time concentrating on the capabilities docs, so I'm a bit out of depth. But here are some things I noticed. Overall, it seems fairly sensible.

File: core/services/ccipcapability/delegate.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+		relayGetter:     relayGetter,
+		capRegistry:     registrySyncer,
+		ds:              ds,
+		keystore:        keystore,
+		peerWrapper:     peerWrapper,
+	}
+}
+
+func (d *Delegate) JobType() job.Type {
+	return job.CCIP
```
Comment (dimkouv) [inline core/services/ccipcapability/delegate.go:67]: Single `job` for both plugins right?

Comment (dimkouv) [inline core/services/ccipcapability/delegate.go:69]: Wondering if the single `job` for both plugins is going to lead to any kind of limitations.

File: core/services/ccipcapability/delegate.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+func (d *Delegate) BeforeJobCreated(job.Job) {
+	// This is only called first time the job is created
+	d.isNewlyCreatedJob = true
```
Comment (dimkouv) [inline core/services/ccipcapability/delegate.go:74]: This seems weird (maybe after it starts being used it becomes clear though): for a newly created delegate `newJob = false`.

File: core/services/ccipcapability/delegate.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+func (d *Delegate) ServicesForSpec(ctx context.Context, spec job.Job) (services []job.ServiceCtx, err error) {
+	// In general there should only be one P2P key but the node may have multiple.
+	// The job spec should specify the correct P2P key to use.
+	peerID, err := p2pkey.MakePeerID(spec.CCIPSpec.P2PKeyID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to make peer ID from provided spec p2p id: %s", spec.CCIPSpec.P2PKeyID)
```
Comment (dimkouv) [inline core/services/ccipcapability/delegate.go]: let's use `fmt.Errorf` everywhere ?

File: core/services/ccipcapability/delegate.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	}
+
+	p2pID, err := d.keystore.P2P().Get(peerID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get all p2p keys")
+	}
+
+	ocrKeys := make(map[chaintype.ChainType]ocr2key.KeyBundle)
+	for chainType, bundleAny := range spec.CCIPSpec.OCRKeyBundleIDs {
```
Comment (dimkouv) [inline core/services/ccipcapability/delegate.go]: Maybe rename `bundleAny` to `bundleIDRaw` ?

File: core/services/ccipcapability/delegate.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to make peer ID from provided spec p2p id: %s", spec.CCIPSpec.P2PKeyID)
+	}
+
+	p2pID, err := d.keystore.P2P().Get(peerID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get all p2p keys")
+	}
+
+	ocrKeys := make(map[chaintype.ChainType]ocr2key.KeyBundle)
```
Comment (dimkouv) [inline core/services/ccipcapability/delegate.go]: Can we put this in a separate function?
```
ocrKeys, err = parseOcrKeys(spec.CCIPSpec.OCRKeyBundleIDs)
```

File: core/services/ccipcapability/delegate.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+	relayers, err := d.relayGetter.GetIDToRelayerMap()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get all relayers")
+	}
+
+	transmitterKeys := make(map[types.RelayID][]string)
+	for relayID := range relayers {
+		switch relayID.Network {
+		case types.NetworkEVM:
```
Comment (dimkouv) [inline core/services/ccipcapability/delegate.go]: To avoid having a huuuge switch case while new chains are getting ID let's improve this a bit.

maybe:
```
transmitterKeysFn = map[networkType]func(...) ([]string, error) {
   types.NetworkEVM: d.evmTransmitterKeys(relayID)
   types.NetworkCosmos: d.cosmosTransmitterKeys(relayID)
}
```


This part is also duplicate
```
transmitterKeys[relayID] = func() (r []string) {
				for _, key := range cosmosKeys {
					r = append(r, key.String())
				}
				return
			}()
```

File: core/services/ccipcapability/delegate.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+					r = append(r, key.String())
+				}
+				return
+			}()
+		default:
+			return nil, errors.Errorf("unsupported network: %s", relayID.Network)
+		}
+	}
+
+	// NOTE: we can use the same DB for all plugin instances,
```
Comment (dimkouv) [inline core/services/ccipcapability/delegate.go:105]: This DB is not used at all, correct?

Comment (dimkouv) [reply core/services/ccipcapability/delegate.go:105]: or is it used internally by ocr?

File: core/services/ccipcapability/launcher/bluegreen.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+	err = multierr.Append(err, c.exec.blue.Shutdown())
+	if c.exec.green != nil {
+		err = multierr.Append(err, c.exec.green.Shutdown())
+	}
+	return err
+}
+
+// NumCommitInstances returns the number of commit OCR instances in the deployment.
+func (c *ccipDeployment) NumCommitInstances() int {
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/bluegreen.go]: Should we change to `HasGreenInstance() bool` ?
It's always going to be `1` or `2`, right?

Comment (dimkouv) [reply core/services/ccipcapability/launcher/bluegreen.go]: That way the conditions will become cleaner
```
&& NumCommitInstances() == 2 {

==>

&& HasGreenInstance()
```

File: core/services/ccipcapability/launcher/diff.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/keystone/generated/keystone_capability_registry"
+	cctypes "github.com/smartcontractkit/chainlink/v2/core/services/ccipcapability/types"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/p2pkey"
+)
+
+type diffResult struct {
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go:13]: Looking at the type I didn't figure out what the `uint32` is, had to look at the code to understand. Maybe a comment?

File: core/services/ccipcapability/launcher/diff.go [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+
+func diff(
+	capabilityVersion,
+	capabilityLabelledName string,
+	oldState,
+	newState cctypes.RegistryState,
+) (diffResult, error) {
+	ccipCapability, err := checkCapabilityPresence(capabilityVersion, capabilityLabelledName, newState)
+	if err != nil {
+		return diffResult{}, err
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go]: wrap errs?

File: core/services/ccipcapability/launcher/diff.go [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+		removed: removed,
+		updated: updated,
+	}, nil
+}
+
+func compareDONs(
+	currCCIPDONs,
+	newCCIPDONs map[uint32]keystone_capability_registry.CapabilityRegistryDONInfo,
+) (
+	added, removed, updated map[uint32]keystone_capability_registry.CapabilityRegistryDONInfo,
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go]: you can return `diffResult` instead of 3 maps

File: core/services/ccipcapability/launcher/diff.go [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+	err error,
+) {
+	added = make(map[uint32]keystone_capability_registry.CapabilityRegistryDONInfo)
+	removed = make(map[uint32]keystone_capability_registry.CapabilityRegistryDONInfo)
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go]: import alias `keystone_capability_registry ==> kcr`
it consumes too much space :D

File: core/services/ccipcapability/launcher/diff.go [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+	updated = make(map[uint32]keystone_capability_registry.CapabilityRegistryDONInfo)
+
+	for id, don := range newCCIPDONs {
+		if currDONState, ok := currCCIPDONs[id]; !ok {
+			// Not in current state, so mark as added.
+			added[id] = don
+		} else {
+			// If its in the current state and the config count for the DON has changed, mark as updated.
+			// Since the registry returns the full state we need to compare the config count.
+			if don.ConfigCount > currDONState.ConfigCount {
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go:67]: `ConfigCount` is `ConfigVersion`, correct?

Comment (dimkouv) [reply core/services/ccipcapability/launcher/diff.go:67]: and for that reason we don't need to check any other field

File: core/services/ccipcapability/launcher/diff.go [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+				updated[id] = don
+			}
+		}
+	}
+
+	for id, don := range currCCIPDONs {
+		if _, ok := newCCIPDONs[id]; !ok {
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go:74]: The name `current` and `new` seems a bit confusing.
Maybe rename to `ccipDonsBefore` `ccipDonsNow`
```
added // now but not before
removed // before but not now
updated // before and now but now is different
```

File: core/services/ccipcapability/launcher/diff.go [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+	state cctypes.RegistryState,
+) (map[uint32]keystone_capability_registry.CapabilityRegistryDONInfo, error) {
+	ccipDONs := make(map[uint32]keystone_capability_registry.CapabilityRegistryDONInfo)
+	for _, don := range state.DONs {
+		// CCIP DONs should only have one capability, CCIP.
+		var found bool
+		for _, donCapabilities := range don.CapabilityConfigurations {
+			if donCapabilities.CapabilityId == hashedCapabilityId(ccipCapability.Version, ccipCapability.LabelledName) {
+				ccipDONs[don.Id] = don
+				found = true
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go:98]: Do you want to `break` after finding it?
Or If a duplicate exists maybe error?

File: core/services/ccipcapability/launcher/diff.go [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+			}
+		}
+		if found && len(don.CapabilityConfigurations) > 1 {
+			return nil, fmt.Errorf("found more than one capability (actual: %d) in the CCIP DON %d",
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go:102]: how do we know that more than one is found?
in general this function is a bit hard to follow what's going on.

```
for each don:
   find the capability registry don info
   if found but don has more than one config -> error
   // if not found but don has more than one config -> ok?
```

File: core/services/ccipcapability/launcher/diff.go [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+		return keystone_capability_registry.CapabilityRegistryCapability{},
+			fmt.Errorf("unable to find capability with name %s and version %s in capability registry state",
+				capabilityLabelledName, capabilityVersion)
+	}
+
+	return ccipCapability, nil
+}
+
+func hashedCapabilityId(capabilityVersion, capabilityLabelledName string) (r [32]byte) {
+	capVersionBytes := []byte(capabilityVersion)
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go:135]: maybe just:
```
return sha256.Sum256([]byte(capabilityVersion + "\n" + capabilityLabelledName))
```

Comment (dimkouv) [reply core/services/ccipcapability/launcher/diff.go:135]: Ohh I guess it should match a hash that we compute onchain.
it is always going to be on ethereum (keccak256), correct?

File: core/services/ccipcapability/launcher/diff.go [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+func isMemberOfDON(don keystone_capability_registry.CapabilityRegistryDONInfo, p2pID p2pkey.KeyV2) bool {
+	var found bool
+	for _, node := range don.NodeP2PIds {
+		if node == p2pID.PeerID() {
+			found = true
+			break
+		}
+	}
+	return found
+}
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/diff.go:165]:
```

	for _, node := range don.NodeP2PIds {
		if node == p2pID.PeerID() {
                         return true
		}
	}
	return false
```

File: core/services/ccipcapability/launcher/launcher.go [hunk 24]
```diff
@@ -0,0 +1,340 @@
...
+	for _, ceDep := range l.dons {
+		err = multierr.Append(err, ceDep.Shutdown())
+	}
+
+	return err
+}
+
+// Start implements job.ServiceCtx.
+func (l *launcher) Start(context.Context) error {
+	l.stopChan = make(chan struct{})
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/launcher.go]: you might want to prevent caller from calling `Start()` twice

File: core/services/ccipcapability/launcher/launcher.go [hunk 25]
```diff
@@ -0,0 +1,340 @@
...
+	// dons is a map of CCIP DON IDs to the OCR instances that are running on them.
+	// we can have up to two OCR instances per CCIP plugin, since we are running two plugins,
+	// thats four OCR instances per CCIP DON maximum.
+	dons map[uint32]*ccipDeployment
+}
+
+// Close implements job.ServiceCtx.
+func (l *launcher) Close() error {
+	// shut down the monitor goroutine.
+	close(l.stopChan)
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/launcher.go]: You might want to prevent `Close()` being called before `Start()`

File: core/services/ccipcapability/launcher/launcher.go [hunk 26]
```diff
@@ -0,0 +1,340 @@
...
+			return
+		case <-ticker.C:
+			if err := l.tick(); err != nil {
+				l.lggr.Errorw("Failed to tick", "err", err)
+			}
+		}
+	}
+}
+
+func (l *launcher) tick() error {
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/launcher.go:113]: maybe rename `tick()` to something more meaningful about what happens inside the func

File: core/services/ccipcapability/launcher/launcher.go [hunk 27]
```diff
@@ -0,0 +1,340 @@
...
+	// if diff is correctly processed, update latest state.
+	l.regState = latestState
+	return nil
+}
+
+// processDiff processes the diff between the current and latest capability registry states.
+// for any added OCR instances, it will launch them.
+// for any removed OCR instances, it will shut them down.
+// for any updated OCR instances, it will restart them with the new configuration.
+func (l *launcher) processDiff(diff diffResult) error {
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/launcher.go:145]: I am wondering how bad will it be if somethings errors in between the operations here.

e.g. we add some dons and then adding one don, errors.
in the next `tick()` `l.regState` was not updated so we start the same dons again. having one don always failing to start leads to hundreds of spawned goroutines.

If something fails between removal it looks fine not sure what happens with updates. Maybe nice to get an extra pair of eyes here.

File: core/services/ccipcapability/launcher/launcher.go [hunk 28]
```diff
@@ -0,0 +1,340 @@
...
+	})
+	errGroup.Go(func() error {
+		return execOracle.Start()
+	})
+	err = errGroup.Wait()
+	if err != nil {
+		// shut down the oracles if we failed to start them.
+		errShutdown := commitOracle.Shutdown()
+		if errShutdown != nil {
+			l.lggr.Errorw("Failed to shutdown commit oracle", "err", err)
```
Comment (dimkouv) [inline core/services/ccipcapability/launcher/launcher.go]: One of this errors will always be logged, depending on which oracle failed to start. Maybe do some better handling to prevent the error from being logged, it might cause confusion.

File: core/services/ccipcapability/launcher/launcher.go [hunk 29]
```diff
@@ -0,0 +1,340 @@
...
+	delete(l.dons, id)
+	return nil
+}
+
+// updateDON handles the case where a DON in the capability registry has received a new configuration.
+// In the case of CCIP, which follows blue-green deployment, we either:
+// 1. Create a new oracle (the green instance) and start it.
+// 2. Shut down the blue instance, making the green instance the new blue instance.
+func (l *launcher) updateDON(don keystone_capability_registry.CapabilityRegistryDONInfo) error {
+	if !isMemberOfDON(don, l.p2pID) {
```
Comment (makramkd) [inline core/services/ccipcapability/launcher/launcher.go:230]: Should also check if we are running instances for this DON

File: core/services/ccipcapability/launcher/launcher.go [hunk 30]
```diff
@@ -0,0 +1,340 @@
...
+		return fmt.Errorf("failed to diff capability registry states: %w", err)
+	}
+
+	err = l.processDiff(diffRes)
+	if err != nil {
+		return fmt.Errorf("failed to process diff: %w", err)
+	}
+
+	// if diff is correctly processed, update latest state.
+	l.regState = latestState
```
Comment (makramkd) [inline core/services/ccipcapability/launcher/launcher.go]: Delete is safe but add/update is not in terms of partially successful addDON/updateDON calls

Comment (makramkd) [inline core/services/ccipcapability/launcher/launcher.go]: Maybe should only start everything if none of the state updates errored?

File: core/services/ccipcapability/launcher/bluegreen.go (hunk 3 above)
Comment (makramkd) [reply core/services/ccipcapability/launcher/bluegreen.go:12]: Eventually it'll switch between LOOPPs, that LOOPP or not factor is hidden behind the OracleCreator interface.

Capabilities are created on the capability registry, DONs that support a capability are also created there, so the capability registry is the source of truth on that.

File: core/services/ccipcapability/delegate.go (hunk 7 above)
Comment (makramkd) [reply core/services/ccipcapability/delegate.go:74]: Yeah this is just standard chainlink job stuff, other jobs do this as well

File: core/services/ccipcapability/delegate.go (hunk 11 above)
Comment (makramkd) [reply core/services/ccipcapability/delegate.go]: Thinking of just axing this and keeping it as EVM for now, wdyt? Can add the other families and refactor later

File: core/services/ccipcapability/delegate.go (hunk 12 above)
Comment (makramkd) [reply core/services/ccipcapability/delegate.go:105]: It is used by OCR, yes

File: core/services/ccipcapability/launcher/diff.go (hunk 18 above)
Comment (makramkd) [reply core/services/ccipcapability/launcher/diff.go:67]: No they are different, config count is how many times a don is configured in the capability registry, config version is an OCR specific config field

File: core/services/ccipcapability/launcher/diff.go (hunk 22 above)
Comment (makramkd) [reply core/services/ccipcapability/launcher/diff.go:135]: Yes this needs to match what is in the onchain registry

File: core/services/ccipcapability/launcher/launcher.go (hunk 24 above)
Comment (makramkd) [reply core/services/ccipcapability/launcher/launcher.go]: Good point, can use `StartOnce`

File: core/services/ccipcapability/launcher/launcher.go (hunk 27 above)
Comment (makramkd) [reply core/services/ccipcapability/launcher/launcher.go:145]: You're right, this code is sketchy.

I think we have a couple of options here, we can:
* try to emulate atomic state updates + service starts. this seems quite hard
* try best effort to start oracles, and update only part of the state where we were successful. seems doable

File: core/services/ccipcapability/delegate.go (hunk 6 above)
Comment (makramkd) [reply core/services/ccipcapability/delegate.go:69]: We will find out ;-)

File: core/services/ccipcapability/launcher/diff.go (hunk 14 above)
Comment (makramkd) [reply core/services/ccipcapability/launcher/diff.go:13]: I can do a type alias for don ID

File: core/services/ccipcapability/launcher/diff.go (hunk 21 above)
Comment (makramkd) [reply core/services/ccipcapability/launcher/diff.go:102]: By checking `len(don.CapabilityConfigurations)`. I'm not sure if we really care about this check, I was just thinking we probably don't want the CCIP don to have more than the CCIP capability 🤔

File: core/services/ccipcapability/launcher/launcher.go (hunk 26 above)
Comment (makramkd) [reply core/services/ccipcapability/launcher/launcher.go:113]: Can't come up with a good name :-) please suggest
